File: pesquisa_operacional/JogoDePoker.java
import java.util.ArrayList;   // Importa a classe ArrayList para criar listas dinâmicas.
import java.util.Collections; // Importa a classe Collections que fornece métodos utilitários para coleções, como embaralhar.
import java.util.List;        // Importa a interface List que representa uma coleção ordenada de elementos.
import java.util.Random;      // Importa a classe Random para gerar números aleatórios (usado para o baralho).
import java.util.Scanner;     // Importa a classe Scanner para obter entrada do usuário (nome dos jogadores, ações).

// Representa um naipe de uma carta
enum Naipe {
    PAUS, // Naipe de Paus
    OUROS, // Naipe de Ouros
    COPAS, // Naipe de Copas
    ESPADAS // Naipe de Espadas
}

// Representa o valor de uma carta
enum Valor {
    DOIS("2"), // Valor 2
    TRES("3"), // Valor 3
    QUATRO("4"), // Valor 4
    CINCO("5"), // Valor 5
    SEIS("6"), // Valor 6
    SETE("7"), // Valor 7
    OITO("8"), // Valor 8
    NOVE("9"), // Valor 9
    DEZ("10"), // Valor 10
    VALETE("J"), // Valor Valete
    DAMA("Q"), // Valor Dama
    REI("K"), // Valor Rei
    AS("A");    // Valor Ás

    private final String simbolo; // Símbolo do valor da carta (para impressão)

    Valor(String simbolo) {
        this.simbolo = simbolo;
    }

    public String getSimbolo() {
        return simbolo;
    }
}

// Representa uma carta de baralho
class Carta {
    private final Naipe naipe; // O naipe da carta
    private final Valor valor; // O valor da carta

    public Carta(Naipe naipe, Valor valor) {
        this.naipe = naipe;
        this.valor = valor;
    }

    public Naipe getNaipe() {
        return naipe;
    }

    public Valor getValor() {
        return valor;
    }

    @Override
    public String toString() {
        return valor.getSimbolo() + " de " + naipe; // Representação da carta como string (ex: "A de ESPADAS")
    }
}

// Representa um jogador
class Jogador {
    private final String nome; // Nome do jogador
    private final List<Carta> mao; // As cartas na mão do jogador
    private int fichas; // Quantidade de fichas que o jogador possui

    public Jogador(String nome, int fichasIniciais) {
        this.nome = nome;
        this.mao = new ArrayList<>();
        this.fichas = fichasIniciais;
    }

    public String getNome() {
        return nome;
    }

    public List<Carta> getMao() {
        return mao;
    }

    public int getFichas() {
        return fichas;
    }

    public void receberCarta(Carta carta) {
        this.mao.add(carta);
    }

    public void apostar(int valor) {
        if (valor <= fichas) {
            fichas -= valor;
        } else {
            System.out.println("Você não tem fichas suficientes para apostar " + valor + ".");
            // Lógica para lidar com aposta inválida (ex: voltar a pedir aposta)
        }
    }

    public void ganharFichas(int valor) {
        fichas += valor;
    }

    public void mostrarMao() {
        System.out.println(nome + " tem na mão: " + mao);
    }

    public void limparMao() {
        this.mao.clear();
    }
}

// Representa o baralho
class Baralho {
    private final List<Carta> cartas; // A lista de cartas no baralho

    public Baralho() {
        this.cartas = criarBaralho();
        embaralhar();
    }

    private List<Carta> criarBaralho() {
        List<Carta> novoBaralho = new ArrayList<>();
        for (Naipe naipe : Naipe.values()) { // Itera por todos os naipes
            for (Valor valor : Valor.values()) { // Itera por todos os valores
                novoBaralho.add(new Carta(naipe, valor)); // Cria uma nova carta e adiciona ao baralho
            }
        }
        return novoBaralho;
    }

    public void embaralhar() {
        Collections.shuffle(cartas); // Embaralha a lista de cartas aleatoriamente
    }

    public Carta distribuirCarta() {
        if (!cartas.isEmpty()) {
            return cartas.remove(0); // Remove e retorna a primeira carta do baralho (topo)
        }
        return null; // Retorna null se o baralho estiver vazio
    }
}

public class JogoDePoker {

    private final List<Jogador> jogadores; // A lista de jogadores participando do jogo
    private final Baralho baralho; // O baralho de cartas usado no jogo
    private final List<Carta> mesa; // As cartas comunitárias na mesa
    private int pote; // O pote de fichas acumuladas para a rodada
    private int apostaAtual; // A aposta atual na rodada de apostas
    private int indiceJogadorAtual; // O índice do jogador cuja vez é apostar

    public JogoDePoker(List<Jogador> jogadores) {
        this.jogadores = jogadores;
        this.baralho = new Baralho();
        this.mesa = new ArrayList<>();
        this.pote = 0;
        this.apostaAtual = 0;
        this.indiceJogadorAtual = 0;
    }

    public void iniciarRodada() {
        System.out.println("\n--- Iniciando nova rodada ---");
        baralho.embaralhar(); // Embaralha o baralho no início de cada rodada
        mesa.clear(); // Limpa as cartas comunitárias da rodada anterior
        pote = 0; // Reseta o pote para zero
        apostaAtual = 0; // Reseta a aposta atual para zero
        indiceJogadorAtual = 0; // Reseta o índice do jogador atual

        // Limpar as mãos dos jogadores para a nova rodada
        for (Jogador jogador : jogadores) {
            jogador.limparMao();
        }

        // Distribuir 2 cartas para cada jogador (mão inicial)
        for (int i = 0; i < 2; i++) {
            for (Jogador jogador : jogadores) {
                Carta carta = baralho.distribuirCarta();
                if (carta != null) {
                    jogador.receberCarta(carta);
                } else {
                    System.out.println("Erro: Baralho acabou durante a distribuição.");
                    return;
                }
            }
        }

        // Mostrar as mãos dos jogadores (apenas para teste - remover em um jogo real)
        for (Jogador jogador : jogadores) {
            jogador.mostrarMao();
        }

        // Fase de apostas pré-flop (antes das cartas comunitárias)
        faseDeApostas();

        // Flop: Distribuir 3 cartas comunitárias na mesa
        if (jogadoresAtivos() > 1) {
            distribuirCartasComunitarias(3);
            System.out.println("Mesa: " + mesa);
            faseDeApostas();
        }

        // Turn: Distribuir a quarta carta comunitária na mesa
        if (jogadoresAtivos() > 1) {
            distribuirCartasComunitarias(1);
            System.out.println("Mesa: " + mesa);
            faseDeApostas();
        }

        // River: Distribuir a quinta e última carta comunitária na mesa
        if (jogadoresAtivos() > 1) {
            distribuirCartasComunitarias(1);
            System.out.println("Mesa: " + mesa);
            faseDeApostas();
        }

        // Showdown: Se mais de um jogador restou, revelar as mãos e determinar o vencedor
        if (jogadoresAtivos() > 1) {
            determinarVencedor();
        } else if (jogadoresAtivos() == 1) {
            // Se apenas um jogador restou, ele vence por desistência dos outros
            for (Jogador jogador : jogadores) {
                if (!jogador.getMao().isEmpty()) {
                    System.out.println(jogador.getNome() + " venceu por desistência.");
                    jogador.ganharFichas(pote);
                    pote = 0;
                    break;
                }
            }
        }
    }

    private void distribuirCartasComunitarias(int numCartas) {
        for (int i = 0; i < numCartas; i++) {
            Carta carta = baralho.distribuirCarta();
            if (carta != null) {
                mesa.add(carta);
            } else {
                System.out.println("Erro: Baralho acabou ao distribuir cartas comunitárias.");
                return;
            }
        }
    }

    private void faseDeApostas() {
        Scanner scanner = new Scanner(System.in);
        int jogadoresQueApostaram = 0; // Contador de jogadores que fizeram alguma ação na rodada de apostas atual
        int ultimaAposta = 0; // A última aposta feita na rodada

        // Listas para controlar o estado da aposta de cada jogador na rodada atual
        List<Boolean> jogadorAgiu = new ArrayList<>(Collections.nCopies(jogadores.size(), false));
        List<Integer> apostaJogador = new ArrayList<>(Collections.nCopies(jogadores.size(), 0));

        int jogadorAtualApostaIndice = indiceJogadorAtual; // Começa a rodada de apostas com o jogador atual

        // Continua a rodada de apostas enquanto houver mais de um jogador ativo e nem todos tenham apostado/igualado
        while (jogadoresAtivos() > 1 && (jogadoresQueApostaram < jogadoresAtivos() || ultimaAposta > 0)) {
            Jogador jogadorAtual = jogadores.get(jogadorAtualApostaIndice % jogadores.size());

            // Se o jogador ainda tem cartas (não desistiu)
            if (!jogadorAtual.getMao().isEmpty()) {
                System.out.println("\nVez de " + jogadorAtual.getNome() + " (fichas: " + jogadorAtual.getFichas() + ")");
                System.out.println("Aposta atual: " + apostaAtual + ", Pote: " + pote);
                System.out.println("Suas cartas: " + jogadorAtual.getMao());

                // Calcula quanto o jogador precisa apostar para cobrir a aposta atual
                int paraCobrir = apostaAtual - apostaJogador.get(jogadorAtualApostaIndice % jogadores.size());

                System.out.println("Opções: (c)heck, (a)postar " + (paraCobrir > 0 ? "(" + paraCobrir + " para cobrir), " : "") + "(f)old");
                String acao = scanner.next().toLowerCase();

                if (acao.equals("c")) { // Check: passar a vez sem apostar (só possível se a aposta atual for 0 ou já tiver igualado)
                    if (paraCobrir == 0 || jogadorAgiu.get(jogadorAtualApostaIndice % jogadores.size())) {
                        jogadorAgiu.set(jogadorAtualApostaIndice % jogadores.size(), true);
                        jogadoresQueApostaram++;
                    } else {
                        System.out.println("Você precisa cobrir a aposta atual para dar check.");
                        continue;
                    }
                } else if (acao.equals("a")) { // Apostar: fazer uma nova aposta ou aumentar a existente
                    System.out.print("Valor da aposta: ");
                    int valorAposta = scanner.nextInt();
                    if (valorAposta >= paraCobrir && jogadorAtual.getFichas() >= valorAposta) {
                        jogadorAtual.apostar(valorAposta);
                        pote += valorAposta;
                        apostaJogador.set(jogadorAtualApostaIndice % jogadores.size(), apostaJogador.get(jogadorAtualApostaIndice % jogadores.size()) + valorAposta);
                        apostaAtual = Math.max(apostaAtual, apostaJogador.get(jogadorAtualApostaIndice % jogadores.size()));
                        ultimaAposta = apostaAtual;
                        jogadorAgiu.set(jogadorAtualApostaIndice % jogadores.size(), true);
                        jogadoresQueApostaram = 1; // Resetar contador, pois houve uma nova aposta
                        // Atualizar jogadores que já apostaram nesta rodada de apostas
                        for (int i = 0; i < jogadorAgiu.size(); i++) {
                            if (i != jogadorAtualApostaIndice % jogadores.size() && apostaJogador.get(i) < apostaAtual && jogadores.get(i).getMao().size() > 0) {
                                jogadorAgiu.set(i, false);
                            }
                        }
                    } else {
                        System.out.println("Aposta inválida. Deve ser no mínimo " + paraCobrir + " e você deve ter fichas suficientes.");
                        continue;
                    }
                } else if (acao.equals("f")) { // Fold: desistir da rodada
                    jogadorAtual.limparMao(); // Remove as cartas da mão do jogador
                    jogadoresQueApostaram++;
                } else {
                    System.out.println("Ação inválida.");
                    continue;
                }
            }

            jogadorAtualApostaIndice++; // Passa a vez para o próximo jogador
        }
    }

    private int jogadoresAtivos() {
        int ativos = 0;
        for (Jogador jogador : jogadores) {
            if (!jogador.getMao().isEmpty()) {
                ativos++;
            }
        }
        return ativos;
    }

    private void determinarVencedor() {
        System.out.println("\n--- Showdown ---");
        List<Jogador> jogadoresAtivos = new ArrayList<>();
        for (Jogador jogador : jogadores) {
            if (!jogador.getMao().isEmpty()) {
                jogador.mostrarMao();
                jogadoresAtivos.add(jogador);
            }
        }
        System.out.println("Mesa: " + mesa);

        if (jogadoresAtivos.size() == 1) {
            Jogador vencedor = jogadoresAtivos.get(0);
            System.out.println(vencedor.getNome() + " é o vencedor! Pote: " + pote);
            vencedor.ganharFichas(pote);
            pote = 0;
            return;
        }

        // TODO: Implementar lógica para avaliar as mãos de poker e determinar o vencedor
        // (straight flush, quadra, full house, flush, straight, trinca, dois pares, um par, carta alta)
        // Por enquanto, o primeiro jogador ativo ganha (para fins de demonstração da estrutura do jogo)
        if (!jogadoresAtivos.isEmpty()) {
            Jogador vencedor = jogadoresAtivos.get(0);
            System.out.println(vencedor.getNome() + " é o vencedor (mão não avaliada)! Pote: " + pote);
            vencedor.ganharFichas(pote);
            pote = 0;
        } else {
            System.out.println("Nenhum jogador ativo para determinar o vencedor.");
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        List<Jogador> jogadores = new ArrayList<>();

        System.out.print("Digite o nome do Jogador 1: ");
        String nomeJogador1 = scanner.nextLine();
        jogadores.add(new Jogador(nomeJogador1, 100)); // Cria o Jogador 1 com 100 fichas

        System.out.print("Digite o nome do Jogador 2 (Computador): ");
        String nomeJogador2 = scanner.nextLine();
        jogadores.add(new Jogador(nomeJogador2, 100)); // Cria o Jogador 2 (Computador) com 100 fichas

        JogoDePoker jogo = new JogoDePoker(jogadores); // Cria uma instância do jogo com os jogadores

        // Continua o jogo enquanto todos os jogadores tiverem fichas suficientes para jogar
        while (jogadores.stream().allMatch(j -> j.getFichas() > 0)) {
            jogo.iniciarRodada(); // Inicia uma nova rodada de poker
            System.out.print("\nPressione Enter para iniciar a próxima rodada...");
            scanner.nextLine(); // Aguarda o jogador pressionar Enter para continuar
        }

        System.out.println("\n--- Fim de Jogo ---");
        for (Jogador jogador : jogadores) {
            System.out.println(jogador.getNome() + " tem " + jogador.getFichas() + " fichas.");
        }
        System.out.println("Obrigado por jogar!");
        scanner.close(); // Fecha o scanner para liberar recursos
    }
}

File: pesquisa_operacional/jogodepoker.py
import random  # Módulo para geração de números aleatórios (embaralhar baralho)

# Definição dos naipes das cartas
NAIPES = ["Paus", "Ouros", "Copas", "Espadas"]

# Definição dos valores das cartas, com seus respectivos símbolos para impressão
VALORES = {
    2: "2", 3: "3", 4: "4", 5: "5", 6: "6", 7: "7", 8: "8", 9: "9",
    10: "10", 11: "J", 12: "Q", 13: "K", 14: "A"
}

class Carta:
    """Representa uma única carta de baralho."""
    def __init__(self, valor, naipe):
        self.valor = valor
        self.naipe = naipe

    def __str__(self):
        """Retorna uma representação legível da carta."""
        return f"{VALORES[self.valor]} de {self.naipe}"

class Baralho:
    """Representa um baralho de 52 cartas."""
    def __init__(self):
        self.cartas = self._criar_baralho()

    def _criar_baralho(self):
        """Cria um baralho completo de 52 cartas."""
        baralho = []
        for naipe in NAIPES:
            for valor in VALORES:
                baralho.append(Carta(valor, naipe))
        return baralho

    def embaralhar(self):
        """Embaralha as cartas do baralho aleatoriamente."""
        random.shuffle(self.cartas)

    def distribuir_carta(self):
        """Distribui a carta do topo do baralho."""
        if self.cartas:
            return self.cartas.pop(0)
        return None  # Retorna None se o baralho estiver vazio

class Jogador:
    """Representa um jogador no jogo de poker."""
    def __init__(self, nome, fichas_iniciais):
        self.nome = nome
        self.mao = []  # Lista para armazenar as cartas na mão do jogador
        self.fichas = fichas_iniciais

    def receber_carta(self, carta):
        """Adiciona uma carta à mão do jogador."""
        self.mao.append(carta)

    def apostar(self, valor):
        """Remove fichas do jogador para fazer uma aposta."""
        if valor <= self.fichas:
            self.fichas -= valor
            return valor
        else:
            print(f"{self.nome} não tem fichas suficientes para apostar {valor}.")
            return 0  # Retorna 0 se a aposta não puder ser feita

    def ganhar_fichas(self, valor):
        """Adiciona fichas ao jogador."""
        self.fichas += valor

    def mostrar_mao(self):
        """Exibe as cartas na mão do jogador."""
        print(f"{self.nome} tem na mão: {[str(carta) for carta in self.mao]}")

    def limpar_mao(self):
        """Remove todas as cartas da mão do jogador."""
        self.mao = []

def fase_de_apostas(jogadores, pote, aposta_atual):
    """Simula uma rodada de apostas."""
    scanner = input  # Usa a função input para obter a ação do jogador
    jogadores_ativos = [jogador for jogador in jogadores if jogador.mao]
    jogadores_que_apostaram = set()
    ultima_aposta = aposta_atual

    # Determina o índice do primeiro jogador a agir (após o dealer na primeira rodada)
    primeiro_a_agir_indice = 0
    if jogadores:
        primeiro_a_agir = jogadores[0] # Assumindo um dealer fixo por simplicidade
        if primeiro_a_agir in jogadores_ativos:
            primeiro_a_agir_indice = jogadores_ativos.index(primeiro_a_agir)

    indice_atual = primeiro_a_agir_indice
    rodada_terminou = False

    while not rodada_terminou and jogadores_ativos:
        jogador_atual = jogadores_ativos[indice_atual % len(jogadores_ativos)]

        if jogador_atual not in jogadores_que_apostaram or ultima_aposta > 0:
            para_cobrir = ultima_aposta - (jogador_atual.aposta_atual if hasattr(jogador_atual, 'aposta_atual') else 0)
            print(f"\nVez de {jogador_atual.nome} (fichas: {jogador_atual.fichas}), Pote: {pote}, Aposta atual: {ultima_aposta}, Precisa cobrir: {para_cobrir}")
            acao = scanner("(c)heck/(a)postar/(f)old: ").lower()

            if acao == 'c':
                if para_cobrir == 0:
                    jogadores_que_apostaram.add(jogador_atual)
                else:
                    print("Você precisa cobrir a aposta para dar check.")
                    continue
            elif acao == 'a':
                valor_aposta = int(scanner("Valor da aposta: "))
                if valor_aposta >= para_cobrir and jogador_atual.fichas >= valor_aposta:
                    aposta_feita = jogador_atual.apostar(valor_aposta)
                    pote += aposta_feita
                    jogador_atual.aposta_atual = (jogador_atual.aposta_atual if hasattr(jogador_atual, 'aposta_atual') else 0) + aposta_feita
                    ultima_aposta = jogador_atual.aposta_atual
                    jogadores_que_apostaram.add(jogador_atual)
                    # Reinicia o conjunto de jogadores que apostaram para a nova aposta
                    jogadores_que_apostaram = {jogador_atual}
                else:
                    print("Aposta inválida.")
                    continue
            elif acao == 'f':
                print(f"{jogador_atual.nome} foldou.")
                jogadores_ativos.remove(jogador_atual)
            else:
                print("Ação inválida.")
                continue

        indice_atual += 1
        if all(jogador in jogadores_que_apostaram for jogador in jogadores_ativos) and all((jogador.aposta_atual if hasattr(jogador, 'aposta_atual') else 0) == ultima_aposta for jogador in jogadores_ativos):
            rodada_terminou = True
        elif not jogadores_ativos:
            rodada_terminou = True

    # Limpa o valor da aposta atual para a próxima rodada de apostas
    for jogador in jogadores_ativos:
        if hasattr(jogador, 'aposta_atual'):
            del jogador.aposta_atual

    return pote, ultima_aposta, jogadores_ativos

def determinar_vencedor(jogadores_ativos, mesa):
    """Determina o vencedor da rodada (implementação básica)."""
    if not jogadores_ativos:
        print("Nenhum jogador ativo para determinar o vencedor.")
        return None

    # Implementação muito simplificada: o jogador com a mão não vazia vence
    vencedor = jogadores_ativos[0]
    print(f"\n--- Showdown ---")
    for jogador in jogadores_ativos:
        jogador.mostrar_mao()
    print(f"Mesa: {[str(carta) for carta in mesa]}")
    print(f"{vencedor.nome} vence a rodada (mão não avaliada).")
    return vencedor

def jogar_rodada(jogadores):
    """Simula uma única rodada de poker."""
    baralho = Baralho()
    baralho.embaralhar()
    mesa = []
    pote = 0
    aposta_atual = 0

    # Limpa as mãos dos jogadores
    for jogador in jogadores:
        jogador.limpar_mao()

    # Distribui 2 cartas para cada jogador
    for _ in range(2):
        for jogador in jogadores:
            carta = baralho.distribuir_carta()
            if carta:
                jogador.receber_carta(carta)
            else:
                print("Erro: Baralho acabou.")
                return

    # Fase de apostas pré-flop
    pote, aposta_atual, jogadores_ativos = fase_de_apostas(jogadores, pote, aposta_atual)
    if len(jogadores_ativos) <= 1:
        return jogadores_ativos[0] if jogadores_ativos else None, pote

    # Flop (3 cartas comunitárias)
    print("\n--- Flop ---")
    for _ in range(3):
        carta = baralho.distribuir_carta()
        if carta:
            mesa.append(carta)
        else:
            print("Erro: Baralho acabou.")
            return
    print(f"Mesa: {[str(carta) for carta in mesa]}")
    pote, aposta_atual, jogadores_ativos = fase_de_apostas(jogadores_ativos, pote, aposta_atual)
    if len(jogadores_ativos) <= 1:
        return jogadores_ativos[0] if jogadores_ativos else None, pote

    # Turn (1 carta comunitária)
    print("\n--- Turn ---")
    carta = baralho.distribuir_carta()
    if carta:
        mesa.append(carta)
    else:
        print("Erro: Baralho acabou.")
        return
    print(f"Mesa: {[str(carta) for carta in mesa]}")
    pote, aposta_atual, jogadores_ativos = fase_de_apostas(jogadores_ativos, pote, aposta_atual)
    if len(jogadores_ativos) <= 1:
        return jogadores_ativos[0] if jogadores_ativos else None, pote

    # River (1 carta comunitária)
    print("\n--- River ---")
    carta = baralho.distribuir_carta()
    if carta:
        mesa.append(carta)
    else:
        print("Erro: Baralho acabou.")
        return
    print(f"Mesa: {[str(carta) for carta in mesa]}")
    pote, aposta_atual, jogadores_ativos = fase_de_apostas(jogadores_ativos, pote, aposta_atual)

    # Determinar o vencedor
    vencedor = determinar_vencedor(jogadores_ativos, mesa)
    return vencedor, pote

def main():
    """Função principal para iniciar o jogo."""
    scanner = input
    num_jogadores = int(scanner("Digite o número de jogadores: "))
    jogadores = []
    for i in range(num_jogadores):
        nome = scanner(f"Digite o nome do Jogador {i+1}: ")
        fichas = 100  # Fichas iniciais para cada jogador
        jogadores.append(Jogador(nome, fichas))

    while any(jogador.fichas > 0 for jogador in jogadores) and len([j for j in jogadores if j.fichas > 0]) > 1:
        vencedor, pote_da_rodada = jogar_rodada(jogadores)
        if vencedor:
            vencedor.ganhar_fichas(pote_da_rodada)
            print(f"{vencedor.nome} ganhou {pote_da_rodada} fichas nesta rodada.")
        else:
            print("Nenhum vencedor nesta rodada.")

        print("\n--- Fim da Rodada ---")
        for jogador in jogadores:
            print(f"{jogador.nome} tem {jogador.fichas} fichas.")

        jogar_novamente = scanner("Jogar outra rodada? (s/n): ").lower()
        if jogar_novamente != 's':
            break

    print("\n--- Fim do Jogo ---")
    for jogador in jogadores:
        print(f"{jogador.nome} terminou com {jogador.fichas} fichas.")

if __name__ == "__main__":
    main()
